### PR TITLE
Add agentic judge audit artifacts

### LINKED ISSUE #722: OpenSearch-VL alignment: add agentic multimodal evaluation suites - Milestone 3: add judge-backed scoring where exact match is insufficient - Unit 1: Add judge prompt snapshots and judge audit JSONL for agentic suites.
## Executable Unit

Add judge prompt snapshots and judge audit JSONL for agentic suites.

## Parent Context

- Direction: `OpenSearch-VL alignment: add agentic multimodal evaluation suites`
- Plan/Milestone: `Milestone 3: add judge-backed scoring where exact match is insufficient`

## Definition Of Done

- The change is scoped to the smallest useful implementation slice.
- The governing spec or plan is created or updated before broad code changes.
- Tests cover the new behavior or contract edge case.
- Metrics and verification output are recorded in the PR evidence.

## Suggested Verification

- `git diff --check`
- Focused pytest or Swift test for the touched code path
- Changed-scope coverage for modified Python or Swift files when measurable
- Any relevant benchmark/evaluation artifact if this changes measured behavior

### PR DESCRIPTION
**Pull request by Zigfreidish**
## Summary
- Closes #722.
- Add audit-only agentic judge prompt snapshot and judge audit JSONL artifacts for evaluation packages that declare `agentic_suite_family`.
- Persist the new artifact paths and prompt metadata in evaluation job params and run evidence without changing scorer behavior or calling a remote judge.
- Document the artifact contract and extend PR-scoped probe coverage for the new evaluation/store paths.

## Plan or Spec
- `docs/plans/2026-05-21-agentic-evaluation-judge-audit.md`
- `docs/benchmark-evaluation-contract.md#agentic-evaluation-judge-audit-artifacts`

## Commands Run
```text
python3 -m json.tool infra/perf/pr_scoped_probes.json >/dev/null && echo json-ok
# json-ok

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" uv run --frozen --project services/mlx-worker-python pytest -q services/mlx-worker-python/tests/test_evaluation_core.py::test_run_local_suite_writes_agentic_judge_prompt_snapshot_and_audit services/mlx-worker-python/tests/test_evaluation_core.py::test_run_local_suite_returns_agentic_judge_artifacts_without_jobs_root services/mlx-worker-python/tests/test_evaluation_store.py::test_persist_result_includes_extra_artifact_paths_in_evidence services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_evaluation_probe_commands_cover_agentic_trajectory_execution services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_evaluation_store_probe_commands_cover_extra_artifact_paths services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_registered_probe_registry_entries_validate_commands_and_watch_globs
# 6 passed in 0.82s

COVERAGE_FILE=/tmp/opensearch-vl-judge-audit-after-merge.coverage PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" uv run --frozen --project services/mlx-worker-python coverage run -m pytest -q services/mlx-worker-python/tests/test_evaluation_core.py::test_run_local_suite_writes_agentic_judge_prompt_snapshot_and_audit services/mlx-worker-python/tests/test_evaluation_core.py::test_run_local_suite_returns_agentic_judge_artifacts_without_jobs_root services/mlx-worker-python/tests/test_evaluation_core.py::test_run_local_suite_persists_agentic_tool_evidence services/mlx-worker-python/tests/test_evaluation_core.py::test_run_local_suite_injects_agentic_tool_trace_before_scoring services/mlx-worker-python/tests/test_evaluation_store.py::test_persist_result_includes_extra_artifact_paths_in_evidence services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_evaluation_probe_commands_cover_agentic_trajectory_execution services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_evaluation_store_probe_commands_cover_extra_artifact_paths services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_registered_probe_registry_entries_validate_commands_and_watch_globs && COVERAGE_FILE=/tmp/opensearch-vl-judge-audit-after-merge.coverage PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" uv run --frozen --project services/mlx-worker-python coverage json -o /tmp/opensearch-vl-judge-audit-after-merge.coverage.json && PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" uv run --frozen --project services/mlx-worker-python python scripts/python_changed_line_coverage.py --coverage-json /tmp/opensearch-vl-judge-audit-after-merge.coverage.json --diff-from origin/main services/mlx-worker-python/worker/engine/evaluation_core.py services/mlx-worker-python/worker/productization/evaluation_store.py services/mlx-worker-python/tests/test_evaluation_core.py services/mlx-worker-python/tests/test_evaluation_store.py services/mlx-worker-python/tests/test_pr_scoped_performance.py
# 8 passed in 0.74s
# TOTAL 100.00% 161/161

make py-test
# 2954 passed, 14 skipped, 2 warnings in 128.41s

pre-commit hook during commit
# make swift-test: passed
# make py-test: 2953 passed, 14 skipped, 2 warnings in 122.60s
# make integration-test: 114 passed, 1 skipped in 310.27s
# PR-scoped performance report: Status ok; changed files 8; selected probes 108; direct/gated probes 9; regressions 0; context regressions 9; verification failures 0
```

## Coverage and Metrics
- Changed-scope coverage: `TOTAL 100.00% 161/161` for the Python files and tests changed in this PR.
- Pre-commit performance report: `/Users/chenyu/Documents/github/melix/.runtime/worktrees/opensearch-vl-judge-audit/.runtime/pre-commit-performance/20260521-113454-2063d57c/report/report.md`.
- PR-scoped performance status: `ok`; direct/gated probes: `9`; direct/gated regressions: `0`; verification failures: `0`; context regressions: `9`.
- Observability mode: `evidence` for evaluation result artifacts only.
- Probe overhead: one prompt snapshot JSONL row per run and one audit JSONL row per selected agentic sample; no production serving request path change and no remote judge call.

## Known Gaps
- No remote judge invocation or judge-backed score change in this slice.
- No strict no-leak enforcement beyond the documented audit artifact boundary; that remains follow-up work under the OpenSearch-VL judge-contract issue family.

## Evidence Checklist
- [x] Relevant plan or spec is identified.
- [x] Behavior changes are reflected in the relevant docs.
- [x] Protocol changes include regenerated generated artifacts.
- [x] Dependency changes include updated lockfiles.
- [x] Relevant tests were run.
- [x] A metrics report is included, or `N/A` is stated explicitly with the reason.
- [x] Observability mode, probe overhead, and any deferred debug or evidence work are recorded, or `N/A` is stated explicitly with the reason.
- [x] Deferred work and known gaps are stated explicitly.

### CODE REVIEW

File: services/mlx-worker-python/worker/engine/evaluation_core.py [hunk 1]
```diff
@@ -2186,6 +2225,166 @@ def _sample_records_for_model(
...
+        allowed_tools = EvaluationCore._string_list(sample.get("allowed_tools"))
+        tool_calls = [dict(call) for call in sample_record.tool_calls]
+        tool_observations = [dict(observation) for observation in sample_record.agentic_tool_observations]
+        user_payload = {
+            "question": question,
+            "expected_answer": expected_answer,
+            "final_answer": sample_record.final_answer,
+            "parse_status": sample_record.parse_status,
+            "scoring_mode": scoring_mode,
+            "typed_score": sample_record.typed_score,
```
Comment (Zigfreidish) [inline services/mlx-worker-python/worker/engine/evaluation_core.py]: **Blocking — label leakage in judge prompt snapshot.**

`typed_score` (the deterministic exact-match verdict) is embedded in the `user_payload` that becomes `messages[1].content`. When a later judge-backed scorer replays this snapshot, it will see the baseline outcome before forming an independent assessment, which defeats the purpose of independent judge evaluation.

The contract doc (`benchmark-evaluation-contract.md`) lists what `messages` must contain: _"question, expected answer, model final answer, parse status, scoring mode, evidence ids, media refs, raw tool calls, and executed tool observations"_ — `typed_score` is explicitly not in that list.

`typed_score` is correctly recorded in the audit row; remove it from `user_payload` here to keep the prompt snapshot judge-agnostic.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: services/mlx-worker-python/worker/engine/evaluation_core.py [hunk 2]
```diff
@@ -2186,6 +2225,166 @@ def _sample_records_for_model(
...
+        dataset_id: str,
+        agentic_suite_family: str,
+        scoring_mode: str,
+        samples: list[dict[str, object]],
+        sample_records: tuple[EvaluationSample, ...],
+    ) -> dict[str, Path]:
+        run_root.mkdir(parents=True, exist_ok=True)
+        snapshot_path = run_root / "agentic-judge-prompt-snapshots.jsonl"
+        audit_path = run_root / "agentic-judge-audit.jsonl"
+        snapshot_rows: list[dict[str, object]] = []
```
Comment (Zigfreidish) [inline services/mlx-worker-python/worker/engine/evaluation_core.py]: Both `snapshot_rows` and `audit_rows` are fully materialized in memory before writing. For large agentic suites this doubles per-sample memory footprint. `_write_jsonl_rows` already accepts an iterable, so consider opening each file once and writing rows incrementally rather than accumulating two full lists. Not blocking for this audit-only slice, but worth addressing before this pattern is reused for larger suites.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: services/mlx-worker-python/worker/engine/evaluation_core.py [hunk 3]
```diff
@@ -2186,6 +2225,166 @@ def _sample_records_for_model(
...
+    def _object_list(value: object) -> list[object]:
+        if not isinstance(value, list):
+            return []
+        return list(value)
+
+    @staticmethod
+    def _string_list(value: object) -> list[str]:
+        if not isinstance(value, list):
+            return []
+        return [item for item in value if isinstance(item, str)]
```
Comment (Zigfreidish) [inline services/mlx-worker-python/worker/engine/evaluation_core.py]: Non-string list items are silently discarded. If a dataset has an `allowed_tools` entry that is a dict (e.g. a dataset authoring bug), the tool will be silently dropped from the snapshot, causing the judge to see a different tool set than the one actually used during evaluation. Since this field is recorded for replay fidelity, consider raising `ValueError` instead of filtering, so dataset validation errors surface early rather than producing a quietly wrong snapshot.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: services/mlx-worker-python/worker/productization/evaluation_store.py [hunk 4]
```diff
@@ -97,6 +98,8 @@ def persist_result(
             "summary_json": summary_path,
             "summary_csv": summary_csv_path,
         }
+        if extra_artifact_paths:
```
Comment (Zigfreidish) [inline services/mlx-worker-python/worker/productization/evaluation_store.py:101]: The `extra_artifact_paths` merge must happen before `build_evaluation_run_evidence` is called so the extra artifact roles appear in `run-evidence.json`. This ordering is correct but fragile — a future refactor that reorders the `persisted` dict construction would silently break evidence inclusion. A short inline comment like `# merge before evidence build so extra roles appear in run-evidence` would make the ordering dependency explicit and prevent silent regressions.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: services/mlx-worker-python/tests/test_evaluation_core.py [hunk 5]
```diff
@@ -865,6 +865,179 @@ def test_run_local_suite_persists_agentic_tool_evidence(tmp_path: Path) -> None:
     assert persisted_samples[0]["agentic_tool_registry"]["toolset_version"] == "melix.agentic_tools.builtin.v1"
 
 
+def test_run_local_suite_writes_agentic_judge_prompt_snapshot_and_audit(
```
Comment (Zigfreidish) [inline services/mlx-worker-python/tests/test_evaluation_core.py:883]: Both new tests use a scripted backend that produces a matching answer, so `typed_score=1.0`, `parse_status="extracted"`, and `failure_stage=""` are always the inputs to the audit row. A failed-sample path (`score=0`, `parse_status="failed"`, non-empty `failure_stage`) is not exercised. Since the audit row exists specifically to record failure context for later judge replay, consider adding a parametrized or second test case with a mismatched answer to confirm those fields are written correctly.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by Zigfreidish — COMMENTED** [review]
The core architecture is sound — artifacts are correctly gated on `agentic_suite_family`, the prompt hash is stable, the no-remote-judge invariant is maintained, changed-scope coverage is 100%, and the probe registry is consistently updated.

**One issue worth fixing before merge:** `typed_score` (the deterministic exact-match verdict) is embedded in `user_payload` inside `messages[1].content` of the judge prompt snapshot. When a later judge-backed scorer replays these snapshots, the judge will see the baseline outcome before forming its own independent assessment — label leakage that undermines the point of independent judge scoring. The contract doc explicitly lists what `messages` must contain and `typed_score` is not in that list. Removing it from `user_payload` in `_agentic_judge_prompt_snapshot_row` fixes this; it already appears correctly in the audit row.

The remaining inline notes (memory accumulation, silent `_string_list` filtering, the `extra_artifact_paths` ordering comment, and the missing failed-sample test) are lower priority but worth addressing before judge-backed scoring builds on top of this artifact format.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: services/mlx-worker-python/worker/engine/evaluation_core.py (hunk 1 above)
Comment (Zigfreidish) [reply services/mlx-worker-python/worker/engine/evaluation_core.py]: Fixed in a0965e89. The prompt snapshot user payload no longer includes typed_score, so the deterministic verdict is only kept in the audit row and is not visible to a future judge replay.

File: services/mlx-worker-python/worker/engine/evaluation_core.py [hunk 6]
```diff
@@ -2186,6 +2225,166 @@ def _sample_records_for_model(
...
+        sample: dict[str, object],
+        sample_record: EvaluationSample,
+    ) -> dict[str, object]:
+        question = EvaluationCore._agentic_sample_question(sample, sample_record)
+        expected_answer = EvaluationCore._target_text_for_sample(sample)
+        evidence_ids = EvaluationCore._object_list(sample.get("evidence_ids"))
+        media_refs = EvaluationCore._object_list(sample.get("media_refs"))
+        allowed_tools = EvaluationCore._string_list(sample.get("allowed_tools"))
+        tool_calls = [dict(call) for call in sample_record.tool_calls]
+        tool_observations = [dict(observation) for observation in sample_record.agentic_tool_observations]
```
[the remark this answers is not in the document]
Comment (Zigfreidish) [reply services/mlx-worker-python/worker/engine/evaluation_core.py]: Fixed in a0965e89. The snapshot serializer now accepts dicts, dataclasses, objects with to_dict(), and key/value iterables, and raises a typed error for non-object payloads. Added coverage in the artifact test.

File: services/mlx-worker-python/worker/engine/evaluation_core.py [hunk 7]
```diff
@@ -2186,6 +2225,166 @@ def _sample_records_for_model(
...
+        if not isinstance(value, list):
+            return []
+        return list(value)
+
+    @staticmethod
+    def _string_list(value: object) -> list[str]:
+        if not isinstance(value, list):
+            return []
+        return [item for item in value if isinstance(item, str)]
+
```
[the remark this answers is not in the document]
Comment (Zigfreidish) [reply services/mlx-worker-python/worker/engine/evaluation_core.py:2424]: Fixed in a0965e89. evidence_ids/media_refs and allowed_tools now accept list or tuple inputs. allowed_tools is strict: non-string entries raise ValueError so replay metadata cannot silently drop malformed tool names.

File: services/mlx-worker-python/worker/engine/evaluation_core.py (hunk 2 above)
Comment (Zigfreidish) [reply services/mlx-worker-python/worker/engine/evaluation_core.py]: Addressed in a0965e89. The artifact writer now opens both JSONL files once and writes snapshot/audit rows incrementally, so it no longer materializes full snapshot_rows and audit_rows lists before writing.

File: services/mlx-worker-python/worker/engine/evaluation_core.py (hunk 3 above)
Comment (Zigfreidish) [reply services/mlx-worker-python/worker/engine/evaluation_core.py]: Fixed in a0965e89. allowed_tools now raises ValueError for non-string entries instead of filtering them out, so malformed dataset metadata fails visibly.

File: services/mlx-worker-python/worker/productization/evaluation_store.py (hunk 4 above)
Comment (Zigfreidish) [reply services/mlx-worker-python/worker/productization/evaluation_store.py:101]: Addressed in a0965e89 with an inline comment documenting that extra artifact paths must be merged before run-evidence generation so their roles are included.

File: services/mlx-worker-python/tests/test_evaluation_core.py (hunk 5 above)
Comment (Zigfreidish) [reply services/mlx-worker-python/tests/test_evaluation_core.py:883]: Addressed in a0965e89. The existing artifact test now also constructs a failed sample audit row and asserts typed_score=0.0, parse_status, failure_stage, failure_reason, and evidence_ids are preserved.
